feat: support multiple sessions - DIA-61984

arththebird opened the pull request:
## Description 

Introduce support for multiple sessions running in parallel. This will primarily be used to support read-only sessions using a different user (connection string) and be able to use AWS Aurora read only replicas. To configure a new separate session, you just have to set a `fastapi_sqla__MY_KEY__sqlalchemy_url` environment variable and it will be configured. See the README for more details about how to use it.

Note that the default session is still configured via `sqlalchemy_url` and all the existing functions default to using the default session if the key is not specified.

## Breaking changes
- Remove support for configuring a separate async session with `async_sqlalchemy_url`. The supported way is now to use an async dialect in the default connection string, or use another session with the proper format.
- The `Session`, `AsyncSession`, `Paginate` and `AsyncPaginate` dependencies are now using the `Annotated` form recommended by FastAPI since version 0.95.0. This means that you'll need to change the usage of the dependencies like: `session: Session = Depends()` -> `session: Session`.
- Drop support for python versions lower than `3.9`. This was already the case since we were only testing with those versions, but the support is now explicitly removed from the `pyproject.toml`.
- The library now requires FastAPI `>= 0.95.1`

## Related

* [DIA-61984]
* #102

## Validation

Default session still works ✅  Tested with [this PR](https://github.com/dialoguemd/mattermerde/pull/188)

Integration tests covers using multiple sessions ✅ 

<!-- 📋 Checklist:
1. Title follows [Commit Convention] and [Code Review guidelines]
   - example: feat(lang): add German language - DIA-12345
2. Relevant labels set
3. Draft PR for WIP
4. Validation notes added
5. Requested from and notified to a team

[Commit Convention]: https://www.notion.so/godialogue/Commit-Convention-84fd9a4c149e48c998d760f1c9176df0
[Code Review guidelines]: https://www.notion.so/godialogue/Code-Review-c5f3fcd185ca49aca73ade497c398fe9  -->


[DIA-61984]: https://dialoguemd.atlassian.net/browse/DIA-61984?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

arththebird commented on README.md:
````diff
@@ -27,7 +27,7 @@ unique `email`:
 
 ```python
 # main.py
-from fastapi import Depends, FastAPI, HTTPException
````
TODO: Document how to use session per key and etc.

arththebird commented on fastapi_sqla/__init__.py:
```diff
@@ -47,18 +53,41 @@
     has_asyncio_support = False
     asyncio_support_err = str(err)
 
+ENGINE_KEYS_REGEX = re.compile(r"fastapi_sqla__(.+)__.+")
```
TODO: Confirm syntax with team

edouardlp replied:
Makes sense to me

arththebird commented on fastapi_sqla/sqla.py:
```diff
@@ -33,27 +33,33 @@
+
+
+def new_engine(*, key: str = "default") -> Engine:
+    envvar_prefix = "sqlalchemy_"
```
TODO: Check if ok to remove ability to override this

arththebird commented on fastapi_sqla/sqla.py:
```diff
@@ -33,27 +33,33 @@
+    __abstract__ = True
+
+
+def new_engine(*, key: str = "default") -> Engine:
```
nit: Make the `default` string a constant and use it everywhere

arththebird commented on pyproject.toml:
```diff
@@ -40,8 +38,8 @@ classifiers = [
 [tool.poetry.dependencies]
-python = "^3.7"
-fastapi = ">=0.61"
+python = "^3.9"
```
We only test with python 3.9+, so I removed support for lower version. (If we don't test with it, we don't support it)

arththebird commented on fastapi_sqla/_pytest_plugin.py:
```diff
@@ -121,18 +120,27 @@ def sqla_transaction(sqla_connection):
-
-    yield fastapi_sqla.sqla._Session(bind=sqla_connection)
+    session = session_factory(bind=sqla_connection)
+    yield session
```
Not absolutely necessary, but it brings it up to par with the async version.

arththebird commented on fastapi_sqla/__init__.py:
```diff
@@ -47,18 +57,48 @@
+                functools.partial(sqla.add_session_to_request, key=key)
+            )
+
+        # TODO: Check if we can get rid of it. I think so
```
TODO

arththebird commented on fastapi_sqla/sqla.py:
```diff
@@ -311,4 +332,5 @@ def paginate(query: DbQuery, scalars=True) -> Page:
 
-Paginate: PaginateDependency = Pagination()
+Paginate = Annotated[PaginateDependency, Depends(Pagination())]
+Session = Annotated[SqlaSession, Depends(SessionDependency())]
```
TBD, but I might be able to make this non breaking by using this form instead. But I feel that since the upgrade path is very simple, it's a good time to introduce a breaking change. We could also get rid of a couple legacy things.

```suggestion
Session: SqlaSession = SessionDependency()
```

arththebird commented on fastapi_sqla/async_pagination.py:
```diff
@@ -1,28 +1,31 @@
 import math
```
TODO: Tests

arththebird commented on fastapi_sqla/async_sqla.py:
```diff
@@ -1,88 +1,79 @@
 import os
```
TODO: Tests

arththebird commented on fastapi_sqla/pagination.py:
```diff
@@ -1,24 +1,24 @@
 import math
```
TODO: Tests

arththebird commented on fastapi_sqla/sqla.py:
```diff
@@ -2,10 +2,10 @@
 import os
```
TODO: Tests

arththebird commented on tests/conftest.py:
```diff
@@ -53,29 +53,28 @@ def is_boto3_installed():
 
 
-@fixture(scope="session")
-def engine(environ):
```
This was moved to the pytest plugin for better consistency

arththebird commented on tests/middleware/conftest.py:
```diff
@@ -0,0 +1,89 @@
+from unittest.mock import Mock
```
The middleware tests were refactored to improve coverage for the async version by adopting a similar pattern as the pagination tests. Consider moving this to its own PR.

arththebird commented on tests/test_async_sqla.py:
```diff
@@ -1,62 +0,0 @@
-
-
-@fixture
-async def startup(environ):
```
The tests related to startup were moved to `test_startup` and the ones related to `open_session` moved to `test_open_session`. Consider moving this to its own PR.

arththebird commented on tests/middleware/test_middleware.py:
```diff
@@ -0,0 +1,94 @@
+    } not in caplog
+
+
+async def test_async_session_dependency_raises_unknown_key(client):
```
```suggestion
async def test_session_dependency_raises_unknown_key(client):
```

hadrien commented:
Just got notified about this hence why I am late to the party. Is there a way to break down these commits in separate PRs?

arththebird commented:
> Just got notified about this hence why I am late to the party. Is there a way to break down these commits in separate PRs?

No worries at all, I just opened the PR for review, so you're not late :)

I already removed a bunch of refactoring related things and moved it to a couple PRs that were merged and [this one](https://github.com/dialoguemd/fastapi-sqla/pull/107) for test refactor. I don't really see a way to split it further because it's kind of all related to the change from a single session to multiple sessions. 

If you see an opportunity to split it, please let me know and it will be my pleasure to update it. FWIW, there's a lot of file that were changed, but not all of them have huge change, it's just so that it uses key based session instead of using the previous `_Session` (or `_AsyncSession`).

hadrien reviewed (commented):
```python
ReadOnlySession = Annotated[SqlaSession, Depends(SessionDependency(key="read_only"))]
```
This is nice. 
1. IIUC, it can be overides in test using something like:
    ```python

    
    def app():
        from ... import ReadOnlySession
        from ...main import app
        app.dependency_override[ReadOnlySession] = Mock()
2. Can we do even simpler? 
    ```python
    from fastqpi_sqla import new_session_dependency
    
    ReadOnlySession = new_session_dependency("read-only")
    ```

hadrien commented:
3. 
    ```python
    ReadOnlySession = Annotated[SqlaSession, Depends(SessionDependency(key="read_only"))]
    ```
    Not sure if this is the case but does this fail only at startup rather than at import?

arththebird commented:
> 3. ```python
>    ReadOnlySession = Annotated[SqlaSession, Depends(SessionDependency(key="read_only"))]
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    Not sure if this is the case but does this fail only at startup rather than at import?

It keeps the same behaviour as before where it only fails at runtime instead of at import. At startup, if you configured a Session with an invalid URL or the db is not accessible, it will fail like before, but for the dependency itself, it only fails at runtime because it's like if you didn't called the `fastapi_sqla.setup` function that adds the middleware that instantiate the session for the request.

arththebird commented:
> ```python
> ReadOnlySession = Annotated[SqlaSession, Depends(SessionDependency(key="read_only"))]
> ```
> 
> This is nice.
> 
> 1. IIUC, it can be overides in test using something like, right?
>    ```python
>    def app():
>        from ... import ReadOnlySession
>        from ...main import app
>        app.dependency_override[ReadOnlySession] = Mock()
>    ```
> 2. Can we do even simpler?
>    ```python
>    from fastapi_sqla import new_session_dependency
>    
>    ReadOnlySession = new_session_dependency("read-only")
>    ```

Good catch for the dependency override in tests! I thought it would work OOTB, but it doesn't. [More details here as to why](https://github.com/tiangolo/fastapi/discussions/9452).

So to have it work, we would need to do something like:
```python
read_only_session_dep = SessionDependency(key="read_only")
ReadOnlySession = Annotated[SqlaSession, Depends(read_only_session_dep)]

def app():
    from ... import read_only_session_dep
    from ...main import app
    app.dependency_override[read_only_session_dep] = Mock()
```

For doing simpler, maybe we could, but it would need more investigation to see whether the type hints are understood by the editors when returning an Annotated type from a function. Also, because of the above, it would only work when we don't want to override it in tests. It definitively is a good suggestion, but I suggest we look into it as a separate story because it's a NTH in my opinion :)

arththebird commented on fastapi_sqla/async_sqla.py:
```diff
@@ -154,3 +139,38 @@ async def get_users(session: fastapi_sqla.AsyncSession = Depends()):
+            raise
+
+
+default_async_session_dep = AsyncSessionDependency()
```
Addresses [this comment](https://github.com/dialoguemd/fastapi-sqla/pull/105#issuecomment-1795317685) for easier overrides

arththebird commented on fastapi_sqla/sqla.py:
```diff
@@ -155,3 +145,36 @@ def get_users(session: fastapi_sqla.Session = Depends()):
+            raise
+
+
+default_session_dep = SessionDependency()
```
Addresses [this comment](https://github.com/dialoguemd/fastapi-sqla/pull/105#issuecomment-1795317685) for easier overrides

edouardlp commented on fastapi_sqla/sqla.py:
```diff
@@ -119,11 +108,12 @@ async def add_session_to_request(request: Request, call_next):
-    async with contextmanager_in_threadpool(open_session()) as session:
-        request.scope[_SESSION_KEY] = session
+    async with contextmanager_in_threadpool(open_session(key)) as session:
+        # TODO: This should use request.state
```
Is this for a future PR?

arththebird replied:
Fixed ✅

edouardlp replied:
Broken you mean 😅 ?

arththebird replied:
oops, I talked too fast 😂

arththebird replied:
Alright, fixed for real this time 😄 That's what happens when I do thee things at the same time 😅

edouardlp reviewed (approved):
Very nice!

hadrien commented:
Bien joué. We have never ever overridden the session dependency anyway so that was a silly comment.

hadrien commented on README.md:
```diff
@@ -153,36 +168,87 @@ class Entity(Base):
 
+# Preferred
+
+read_only_session_dep = SessionDependency(key="read_only")
+ReadOnlySession = Annotated[SqlaSession, Depends(read_only_session_dep)]
+
+async_read_only_session_dep = AsyncSessionDependency(key="read_only")
+AsyncReadOnlySession = Annotated[
+    SqlaAsyncSession, Depends(async_read_only_session_dep)
+]
```
I'd get back to how it was documented without separating the dep from annotation. Seems cleaner to me. And sorry for suggesting it had to be overridden in tests... YAGNI
